src/main/python/mmlspark/cyber/anomaly/collaborative_filtering.py

```diff
@@ -0,0 +1,880 @@
+
+        return self.user_feature_vector_mapping_df.select(
+            self.tenant_col, self.user_col
+        ).distinct().count() == self.user_feature_vector_mapping_df.count()
+
+    def _check_res_mapping(self):
+        field_map = {ff.name: ff for ff in self.res_feature_vector_mapping_df.schema.fields}
+
+        assert field_map.get(self.tenant_col) is not None, field_map
+        assert field_map.get(self.res_col) is not None
+
+        return self.res_feature_vector_mapping_df.select(
+            self.tenant_col, self.res_col
+        ).distinct().count() == self.res_feature_vector_mapping_df.count()
+
+
+class AccessAnomalyModel(Transformer):
+    outputCol = Param(
+        Params._dummy(),
+        "outputCol",
+        "The name of the output column representing the calculated anomaly score. "
+        "Values will be between (-inf, +inf) with an estimated mean of 0.0 and standard deviation of 1.0. "
+    )
+
+    """
+    A pyspark.ml.Transformer model that can predict anomaly scores for user, resource access pairs
+    """
+    def __init__(self, user_res_feature_vector_mapping: UserResourceFeatureVectorMapping, output_col: str):
+        super().__init__()
+        self.user_res_feature_vector_mapping = user_res_feature_vector_mapping
+
+        has_user2component_mappings = self.user_res_feature_vector_mapping.user2component_mappings_df is not None
+        has_res2component_mappings = self.user_res_feature_vector_mapping.res2component_mappings_df is not None
+
+        assert has_user2component_mappings == has_res2component_mappings
+
+        self.has_components = has_user2component_mappings and has_res2component_mappings
+        self.preserve_history = True
+
+        if self.has_components:
+            self._user_mapping_df = self.user_res_feature_vector_mapping.user_feature_vector_mapping_df.join(
+                self.user_res_feature_vector_mapping.user2component_mappings_df, [self.tenant_col, self.user_col]
+            ).select(
+                self.tenant_col,
+                self.user_col,
+                self.user_vec_col,
+                f.col('component').alias('user_component')
+            ).cache()
+
+            self._res_mapping_df = self.user_res_feature_vector_mapping.res_feature_vector_mapping_df.join(
+                self.user_res_feature_vector_mapping.res2component_mappings_df, [self.tenant_col, self.res_col]
+            ).select(
+                self.tenant_col,
+                self.res_col,
+                self.res_vec_col,
+                f.col('component').alias('res_component')
+            ).cache()
+        else:
+            self._user_mapping_df = self.user_res_feature_vector_mapping.user_feature_vector_mapping_df
+            self._res_mapping_df = self.user_res_feature_vector_mapping.res_feature_vector_mapping_df
+
+        spark_utils.ExplainBuilder.build(self, outputCol=output_col)
+
+    @property
+    def tenant_col(self):
```

Comment:
can you use sparkML idiomatic getters and setters (camelCase)

Reply:
Will change param names in ctor to camelCase

Reply:
changed all the param names in the ctors to be camelCase as agreed